### PR TITLE
[TH2-1005]

### PR DESCRIPTION
**Pull request by std000**

### CODE REVIEW

**Review by yarchiksmith — DISMISSED** [review]

File: src/main/java/com/exactprosystems/remotehand/windows/actions/SwitchActiveWindow.java [hunk 1]
```diff
@@ -36,14 +38,26 @@
...
+			ACCESSIBILITY_ID_PARAM = "accessibilityid",
+			MAX_TIMEOUT_PARAM = "maxtimeout";
 	private static final String HANDLE_ATTRIBUTE = "NativeWindowHandle";
 
 	private static final Logger loggerInstance = LoggerFactory.getLogger(SwitchActiveWindow.class);
 
 	@Override
 	public String run(WindowsDriverWrapper driverWrapper, Map<String, String> params, WindowsSessionContext.CachedWebElements cachedWebElements) throws ScriptExecuteException {
 
+		// Action is best usage for login windows, where login window runs under the same process that
```
Comment (Vladimir-Panarin-Exactpro) [inline src/main/java/com/exactprosystems/remotehand/windows/actions/SwitchActiveWindow.java]: replace the comment with:
// Action is useful for login windows, where login window belongs to the same process as
// main application. Action is able to catch required window even if there are multiple windows opened
// that satisfy chosen criteria.

File: src/main/java/com/exactprosystems/remotehand/windows/actions/SwitchActiveWindow.java [hunk 2]
```diff
@@ -55,53 +69,77 @@ public String run(WindowsDriverWrapper driverWrapper, Map<String, String> params
...
+			do {
+				if (!firstIt) {
+					Thread.sleep(1000);
+				}
 
-		WindowsDriver<?> root = driverWrapper.getOrCreateRootDriver();
-		Set<String> allWindowHandles = findWindows(root, targetWindowMatcher, byName);
+				try {
+					if (isCurrentWindowExpected(driver, targetWindowMatcher, byName)) {
+						this.logger.debug("Current window has same title that expected");
```
Comment (Vladimir-Panarin-Exactpro) [inline src/main/java/com/exactprosystems/remotehand/windows/actions/SwitchActiveWindow.java]: Current window has the same title as expected one

File: src/main/java/com/exactprosystems/remotehand/windows/actions/SwitchActiveWindow.java [hunk 3]
```diff
@@ -55,53 +69,77 @@ public String run(WindowsDriverWrapper driverWrapper, Map<String, String> params
...
+					} catch (NoSuchWindowException e) {
+						logger.debug("Window handle: {} title: {} was closed", handle, title);
+					}
+				}
+
+				firstIt = false;
+			} while (maxTimeout >= 0 && (startTime + maxTimeout) >= System.currentTimeMillis());
+			
+		} catch (InterruptedException e) {
+			logger.warn("Interrupted");
```
Comment (Vladimir-Panarin-Exactpro) [inline src/main/java/com/exactprosystems/remotehand/windows/actions/SwitchActiveWindow.java]: Add exception to log to see where the interruption occurred.

File: src/main/java/com/exactprosystems/remotehand/windows/actions/SwitchActiveWindow.java [hunk 4]
```diff
@@ -55,53 +69,77 @@ public String run(WindowsDriverWrapper driverWrapper, Map<String, String> params
...
-			WebElement window = (WebElement) w;
-			String handle = window.getAttribute(HANDLE_ATTRIBUTE);
-			String handleHex = nativeWindowHandleToHex(handle);
-			handles.add(handleHex);
-			logger.debug("{}={}, hex={}", HANDLE_ATTRIBUTE, handle, handleHex);
+				String currentHandle;
+				try {
+					currentHandle = driver.getWindowHandle();
+				} catch (NoSuchWindowException e) {
+					currentHandle = null;
```
Comment (Vladimir-Panarin-Exactpro) [inline src/main/java/com/exactprosystems/remotehand/windows/actions/SwitchActiveWindow.java:92]: Maybe, print something to log to highlight that exception occurred and was suppressed?

File: .github/workflows/dev-java-publish-bintray.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
+name: Dev build and publish Java distributions to Bintray
```
Comment (Vladimir-Panarin-Exactpro) [inline .github/workflows/dev-java-publish-bintray.yml]: Is it expected to have this file here?

**Review by Vladimir-Panarin-Exactpro — CHANGES_REQUESTED** [review]

File: .github/workflows/dev-java-publish-bintray.yml (hunk 5 above)
Comment (std000) [reply .github/workflows/dev-java-publish-bintray.yml]: Yes, but there are many commits there, so I'm going to squash them and its file commit message will be lost. I deleted this one.

**Review by Vladimir-Panarin-Exactpro — DISMISSED** [review]
Better change commit message to reflect that new action was added **and** existing one was changed.

**Review by Vladimir-Panarin-Exactpro — APPROVED** [review]

**Pull request merged**